### PR TITLE
Add support for whitespace inside build args

### PR DESCRIPTION
**Pull request by antdking**
Utilise a bash array to expand the parameters correctly for the img command.

We've been facing an issue where if we have a build argument that contains whitespace, `img` will think the second word is the context.

```sh
$ build_args="--build-arg VAR=some var"
$ img build --no-console -t image:latest $build_args .
... 
failed to solve: failed to resolve dockerfile: error from sender: failed to resolve : lstat var: no such file or directory
```

Changing to a bash array allows us to do parameter expansion:

```sh
$ build_args=( '--build-arg' 'VAR=some var' )
$ img build --no-console -t image:latest "${build_args[@]}" .
# builds as expected with `VAR` set to `some var`
```

### CODE REVIEW

**Comment by vito** [comment]
makes sense! would you mind signing the commit to make the DCO bot happy?: https://github.com/concourse/builder-task/pull/23/checks?check_run_id=126348364

**Comment by antdking** [comment]
Hi @vito, I've rebased the PR to signoff the commits.
I've also fixed whitespace when using `build_args_file`.

This introduce a small behaviour change, so `BUILD_ARG_*` and `BUILD_ARGS_FILE` can be used at the same time.
Let me know if I should change it back to the previous behaviour where `BUILD_ARG_*` is ignored when `BUILD_ARS_FILE` is set

File: build [hunk 1]
```diff
@@ -97,25 +95,26 @@ elif [ -n "$TARGET" ]; then
 fi
 
 
-build_args=""
+build_args=()
+
+while read -r line && test "${line}"; do
```
Comment (vito) [inline build]: The `test` here seems to be missing an operator (`-n`?)

**Review by vito — CHANGES_REQUESTED** [review]
Thanks! I like the idea of supporting both. Noticed one small problem though.

File: build (hunk 1 above)
Comment (antdking) [reply build]: this isn't required, however I've added it to be explicit we're testing for string length:

```sh
$ test ""; echo $?
1
$ test "contents"; echo $?
0
```

Comment (vito) [reply build]: I was more worried about the contents of `${line}` being interpreted by `test`, e.g. if one of the values was a flag. Thanks for the fix! :+1:

File: build [hunk 2]
```diff
@@ -97,25 +95,26 @@ elif [ -n "$TARGET" ]; then
...
+
 if [ -n "$BUILD_ARGS_FILE" ]; then
   if [ ! -f "$BUILD_ARGS_FILE" ]; then
     echo "build_args file '$BUILD_ARGS_FILE' does not exist"
     exit 1
   fi
 
-  expanded_build_args=()
-  while IFS= read -r line || [ -n "$line" ];
+  while IFS= read -r line && test -n "${line}" || [ -n "$line" ];
```
Comment (vito) [inline build:111]: `test -n "${line}" || [ -n "$line" ]` are both the same thing (`[` is an alias to `test`):

```
alex@wx020:~/workspace/builder-task$ help [
[: [ arg... ]
    Evaluate conditional expression.
    
    This is a synonym for the "test" builtin, but the last argument must
    be a literal `]', to match the opening `['.
```

**Review by vito — CHANGES_REQUESTED** [review]
Another quick question/nit regarding `test` usage. :stuck_out_tongue: Sorry.

**Comment by vito** [comment]
I'm working on a rewrite in Go that will, among other things, fix these pesky whitespace/scripting issues. Thanks for the PR though!

**Pull request closed without merging**